src/api/index.js

```diff
@@ -1,104 +1,87 @@
-        response = service.getLocationByFSID(type, this.params);
-      }
-      if (this.lookupType === ADDRESS) {
-        response = service.getLocationByAddress(type, this.params);
-      }
-      if (this.lookupType === COORDINATE) {
-        response = service.getLocationByLatLng(type, this.params);
-      }
-    } else {
-      throw new Error('Must provide a service');
+  callService(service, detailLevel, serviceParams) {
+    assert(detailLevel, 'Detail level is required');
+
+    // find mapping for the endpoint
+    const endpointMapping = mapping(service, detailLevel);
+
+    assert(endpointMapping, `Internal error: cannot find mapping for ${service}/${detailLevel}`);
+    assert(
+      !endpointMapping.needsLocation
+      || (endpointMapping.needsLocation && localities.includes(this._locationType)),
+      `Please set lookup parameters prior to calling service "${service}"`,
+    );
+
+    // validate there's no unknonw parameters
+    const params = serviceParams || {};
+    const allowedParameters = endpointMapping.allowedParameters || [];
+    assert(
+      _.difference(Object.keys(params), allowedParameters).length <= 0,
+      `Unknown parameter(s): ${_.difference(Object.keys(params), allowedParameters)}`,
+    );
+
+    // if not location lookup then check lookup type
+    let lookupType = this._lookupType;
+    if (!endpointMapping.needsLocation) {
+      assert(_.intersection(Object.keys(params), allowedParameters).length > 0, `Service ${service}/${detailLevel} parameter(s) are missing. Required: ${allowedParameters.join(', ')}`);
+      [lookupType] = endpointMapping.allowedParameters;
     }
 
-    return response;
+    // get response
+    return this._resolver.getServiceResponse(
+      endpointMapping,
+      _.merge(this._lookupParams || {}, params),
+      this._locationType,
+      lookupType,
+    );
   }
 
-  location(service) {
-    // service is a str mapping to api location services
-
-    if (!TYPES.includes(this.type)) {
-      throw new Error('Must provide a valid location type');
-    }
-
-    return this.fetchData(this.type, service);
+  static checkLookupParams(params) {
+    assert(params, 'Missing lookup parameters');
+    assert(typeof params === 'object', 'Lookup parameters need to be in an object');
+    assert(_.difference(Object.keys(params), allowedLookupParameters).length <= 0, `Unknown parameter(s): ${_.difference(Object.keys(params), allowedLookupParameters)}`);
   }
 
-  probability(service) {
-    if (!TYPES.includes(this.type)) {
-      throw new Error('Must provide a valid location type');
+  setLookupType(params) {
+    Api.checkLookupParams(params);
+
+    if (params.fsid) {
+      this._lookupType = FSID;
+    } else if (params.address) {
+      this._lookupType = ADDRESS;
+    } else if (params[LAT] || params[LNG]) {
+      assert(params[LAT] && params[LNG], 'Must provide both latitude and longitude for coordinate lookup');
+      this._lookupType = COORDINATE;
```

Comment:
What happens if someone provides all 3?

Reply:
good question. We have these options:
- throw an error and make them choose
- use the first found and ignore the rest of parameters

which one you think is better? I would vote for the first one.